lib/src/barback/dartdevc/dartdevc_environment.dart

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:barback/barback.dart';
+import 'package:cli_util/cli_util.dart' as cli_util;
+import 'package:path/path.dart' as p;
+
+import 'dartdevc.dart';
+import 'module.dart';
+import 'module_computer.dart';
+import 'module_reader.dart';
+import 'scratch_space.dart';
+import 'summaries.dart';
+
+import '../../io.dart';
+import '../../log.dart' as log;
+import '../../package_graph.dart';
+
+/// Handles running dartdevc on top of a [Barback] instance.
+///
+/// You must call [invalidatePackage] any time a package is updated, since
+/// barback doesn't provide a mechanism to tell you which files have changed.
+class DartDevcEnvironment {
+  final _AssetCache _assetCache;
+  final Barback _barback;
+  final Map<String, String> _environmentConstants;
+  final BarbackMode _mode;
+  ModuleReader _moduleReader;
+  final PackageGraph _packageGraph;
+  ScratchSpace _scratchSpace;
+
+  DartDevcEnvironment(
+      this._barback, this._mode, this._environmentConstants, this._packageGraph)
+      : _assetCache = new _AssetCache(_packageGraph) {
+    _moduleReader = new ModuleReader(_readModule);
+    _scratchSpace = new ScratchSpace(_getAsset);
+  }
+
+  /// Deletes the [_scratchSpace].
+  Future cleanup() => _scratchSpace.delete();
+
+  /// Builds all dartdevc files required for all app entrypoints in
+  /// [inputAssets].
+  ///
+  /// Returns only the `.js` files which are required to load the apps.
+  Future<AssetSet> doFullBuild(AssetSet inputAssets,
+      {logError(String message)}) async {
+    var modulesToBuild = new Set<ModuleId>();
+    var jsAssets = new AssetSet();
+    for (var asset in inputAssets) {
+      if (asset.id.package != _packageGraph.entrypoint.root.name) continue;
+      if (asset.id.extension != '.dart') continue;
+      // We only care about real entrypoint modules, we collect those and all
+      // their transitive deps.
+      if (!await isAppEntryPoint(asset.id, _barback.getAssetById)) continue;
+      // Build the entrypoint js files, and collect the set of transitive
+      // modules that are required (will be built later).
+      var futureAssets =
+          _buildAsset(asset.id.addExtension('.js'), logError: logError);
+      jsAssets.addAll((await Future.wait(futureAssets.values))
```

Comment:
The precedence of `await` is a little hard to read. Maybe instead of parenthesizing, hoist it out into a local?

```dart
var assets = await Future.wait(futureAssets.values);
jsAssets.addAll(assets.where(...));
```

?